### ISSUE TITLE
RFE: make portabled profiles available to all units

### ISSUE THREAD

**Issue by damnms** [systemd/systemd#16511]
**Is your feature request related to a problem? Please describe.**
I have a couple of services which i'd like to "secure". So i thought i use systemd to make them secure because its nice and easy to move that to another hardware (override.conf, thanks!). 
But: when i for example check the acct.service (systemd-analyze security acct.service), it has more than 70(!) configuration flags that i have to manually define in the override.conf to be disabled. Thats ridiculous! And if there are new configuration options, they are not noticed by anyone. 

**Describe the solution you'd like**
I would like to have something like (in override.conf)
MaximumSecurity=yes
MaximumSecurityIgnore=DynamicUser
which leads to everything set to no/denied or like IPAddressDeny to true, except DynamicUser which is not activated.
And when i start that service, i can see if it works or not. If it does not work, i add the required flag to MaximumSecurityIgnore

**Describe alternatives you've considered**
no idea what else could be an alternative. maybe use SELinux

**Comment by DemiMarie** [comment]
One difficulty is that some security options (like `RootDirectory=`) have no reasonable defaults.

**Comment by damnms** [comment]
thats true, maybe this needs some special investigation and decisions. 
i would already be happy if at least 50% of the flags would be set by something like MaxSecurity=yes ;)

**Comment by swestlake** [comment]
I don't think that is possible because systemd can't predict whether a service would need to perform write operations.. The systemd units are generally configured by package maintainers but they can be modified by overriding them in /etc/systemd/system/my_service_copied_here.service, instead of /etc/systemd/system/servicename.conf.d/override.conf. As a long time user I never have had to use override.conf, and prefer to just comment out settings to have resources set to their default, or simply start from scratch and just contain the simple enough settings for the system getting maintained.. If you're unsetting most of the things in an override.conf, then imho you should instead be using .service entries..

**Comment by damnms** [comment]
i am not that familiar with systemd. i am just a hobby admin of a mail server and i have a couple of services which are reported by lynis, so i thought i take a deeper look at that.

> because systemd can't predict whether a service would need to perform write operations
thats not the job of systemd and i never asked for something like that. i want to be able to do that by myself. i can do it by myself, when i add all those 70 flags to an override.conf, but thats really annoying.

> The systemd units are generally configured by package maintainers
i have 2 services created by myself and also there, i have to configure all those 70 flags 

> As a long time user I never have had to use override.conf
i also never used override.conf, but i want to separate between system configuration and user configuration. i also do not touch /usr/local/phpmyadmin/config.inc.php because when an update comes out for phpmyadmin, i lose the configuration (just a sample, i do not use phpmyadmin). 

> and prefer to just comment out settings to have resources set to their default
a default setting should be (imo) the most secure one. that would be fine if that would be the case, but when i take a look at systgemd-analyze security, i see a lot of 8.x and 9.x ratings. thats not good.

> If you're unsetting most of the things in an override.conf, then imho you should instead be using .service entries..
what do you mean exactly? i dont understand that.

**Comment by boucman** [comment]
you should look at the "profiles" included with portable services... they are upstream-maintained drop-ins with reasonable defaults for most security features...

**Comment by damnms** [comment]
@boucman could you maybe post a link to "systemd profiles with portable services" ? i did not find any reasonable on a google search, thanks!

**Comment by keszybz** [comment]
> you should look at the "profiles" included with portable services... they are upstream-maintained drop-ins with reasonable defaults for most security features...

Yeah, we should make profiles accessible to all units.

**Comment by bluca** [comment]
> @boucman could you maybe post a link to "systemd profiles with portable services" ? i did not find any reasonable info on a google search, thanks!
> 
> ah, i guess now i understood what you mean.
> portable services are some kind of containers that run with systemd. and the configuration could be copied from them?
> ok, thats a way to go but... it still feels odd to me. because, they normally should also configure all those 70+ flags, shouldnt they?
> and that also works only for services that are maintained. so every service file i create by myself has to be configured with those 70+ flags.
> 
> is there a repository containing those profiles for systemd portable services? i searched only a few seconds, maybe minutes, but did not find something like e.g. dockerhub where i can copy those files.

@damnms you can find them here: https://github.com/systemd/systemd/tree/master/src/portable/profile

Having security profiles generally available sounds like a good idea, I think the topic of having generic security profiles that all units can use came up during the FOSDEM 2020 talk on this topic

**Comment by damnms** [comment]
Oh that would be great if there is something like:
Profile=scrict 

That would also address the issue: if there are new options, i dont have to add them to all my override.conf's

**Comment by boucman** [comment]
profiles are already drop-ins... so you just need to symlink them next to your overrides.conf and they will be taken into account

when you upgrade systemd, new versions will magically come with new security features added... which is a risk by itself since it might block things that your application previously used. But that's a risk you have to wieght as an admin...

**Comment by keszybz** [comment]
Symlinking is an option, but I think having this integrated as an option would be nicer.

**Comment by boucman** [comment]
I suggested symlinks because that's how it's done by portablectl IIRC, but yes. proper integration would be nice.

**Comment by swestlake** [comment]
> i also do not touch /usr/local/phpmyadmin/config.inc.php because when an update comes out for phpmyadmin, i lose the configuration (just a sample, i do not use phpmyadmin).

https://docs.phpmyadmin.net/en/latest/config.html
"Some distributions (eg. Debian or Ubuntu) store config.inc.php in /etc/phpmyadmin instead of within phpMyAdmin sources."

>  i can do it by myself, when i add all those 70 flags to an override.conf, but thats really annoying.

.. and that's why there are other ways to create overrides.

You lose system settings when you don't edit them in /etc.  If you do lose settings for a package, then you should file a bugreport to that package maintainer.

You should never lose any settings whatsoever with package upgrades.   To give me an example as a case tells me you must be editing things at their wrong location.

> i am not that familiar with systemd.

obviously.. more so on where to edit configurations where you think you need to edit 70 things and there is no other way to do it. Nobody is going to take you by the hand to tell you how to read documentation, it's your job as an administrator to read up into and verify what directives make most sense for the unit you're working on.

I still say it is very impractical, because a quick "toggle" only minimizes the "attack surface" for trying to undo that "global" switch so I say it will never happen.  Having profiles makes more sense, but users can already do that for years with SeLinux, Apparmor and other things.  I don't know about Systemd profiles, and I won't pretend that I do, but there are many things Systemd doesn't do and is not the "magic" bullet for all security.

Having a "maximum" setting is not going to get rid of 0-day CVE reports. There is no "magic" bullet, so keep dreaming because there will always be an exploit coming out the next day. 

When you tell me you lose settings when you perform package upgrades, you're not just new to systemd, you're also new to Linux as a whole ecosphere.  So in a way people online reading your comments might think that they should worry about configuration changes --- they shouldn't. 

Try being a little more realistic --- you don't need to keep ranting about keep "needing" 70 things when there are "easier" and more practical ways about it.

Also try learning about how to maintain configuration files more adequately, whether it is php, systemd, etc.. so on.  Your "php.ini" settings in specific get stored in newer paths --> /etc/php/version_number/  --- You're telling me you need a front-end for php is not at all too inviting when you succumb to 70 changes needed in order to make systemd "more secure".

**Comment by dcousens** [comment]
If a concern is that new flags might break existing services, I wouldn't be against the profiles being akin to `strict-20220524` or some kind of versioning.

Then you could configure your profile to when you last checked and easily review changelogs to determine when to bmp your profile to the newest version.

**Comment by DemiMarie** [comment]
@krishjainx that would be amazing but I do not have the time to work on this right now.
